### ISSUE TITLE
proto-gen-go: mixed packages when some files lack go_package

### ISSUE THREAD

**Issue by aronatkins** [golang/protobuf#984]
Trying to build Go definitions for github.com/tensorflow/tensorflow (off the r1.15.0 tag, if it matters):

```bash
protoc \
       -I./github.com/tensorflow/tensorflow \
       --go_out=plugins=grpc:./ \
       ./github.com/tensorflow/tensorflow/tensorflow/core/framework/*.proto
protoc \
       -I./github.com/tensorflow/tensorflow \
       --go_out=plugins=grpc:./ \
       ./github.com/tensorflow/tensorflow/tensorflow/core/protobuf/{saver,saved_model,struct,trackable_object_graph,saved_object_graph,meta_graph}.proto
```

gives the error:
```
2019/11/07 14:43:01 protoc-gen-go: error:inconsistent package import paths: "github.com/tensorflow/tensorflow/tensorflow/go/core/protobuf", "tensorflow/core/protobuf"
```

If I try using loads of `M` arguments, the error goes away, but I get files written both to `github.com/tensorflow/tensorflow/tensorflow/go/core/protobuf` and `tensorflow/core/protobuf`.

```bash
protoc \
       -I./github.com/tensorflow/tensorflow \
       --go_out=plugins=grpc,Mtensorflow/core/protobuf/saved_object_graph.proto=github.com/tensorflow/tensorflow/tensorflow/go/core/protobuf,Mtensorflow/core/protobuf/struct.proto=github.com/tensorflow/tensorflow/tensorflow/go/core/protobuf,Mtensorflow/core/protobuf/trackable_object_graph.proto=github.com/tensorflow/tensorflow/tensorflow/go/core/protobuf:./ \
       ./github.com/tensorflow/tensorflow/tensorflow/core/protobuf/{saver,saved_model,struct,trackable_object_graph,saved_object_graph,meta_graph}.proto
```

That leads to a Go compilation failure:
```
src/github.com/tensorflow/tensorflow/tensorflow/go/core/protobuf/meta_graph.pb.go:54:24: undefined: SavedObjectGraph
```

That type was written to the `tensorflow/core/protobuf/saved_object_graph.pb.go` file with an import back into `github.com/tensorflow/tensorflow/tensorflow/go/core/framework`.

If I add a `go_package` to these three files, everything is emitted into `github.com/tensorflow/tensorflow/tensorflow/go/core`.
*   tensorflow/core/protobuf/saved_object_graph.proto
*   tensorflow/core/protobuf/struct.proto
*   tensorflow/core/protobuf/trackable_object_graph.proto

```
option go_package = "github.com/tensorflow/tensorflow/tensorflow/go/core/protobuf";
```

My question is: What is the recommended way to generate all the Go files into one package hierarchy? Should I need to modify the source? I'm unclear if there is a problem with my `protoc` command, with `proto-gen-go`, or with the TensorFlow code.

**Comment by neild** [comment]
tl;dr, I think the problem is with the TensorFlow code.

"Inconsistent package import paths" is because `protoc-gen-go` wants to only compile one Go package at a time. This limitation will go away in the nearish future, but the simplest workaround right now when you do want to generate code in more than one Go package is to compile one `.proto` file at a time.

From a quick look here, however, you actually do want all generated code in the same package. The problem is that three of the `.proto` files are missing an `option go_package`. The best fix is to add that option to those files; the fact that it's missing is probably an oversight in the TensorFlow code.

The recommended way to generate Go proto files is to put `option go_package = "example.com/full/import/path"` in each file. You can also pass lots of `M` arguments, but that's really only intended for build tooling like Bazel.

The quick explanation as for why is that `protoc-gen-go` needs to know the Go package for each file. For example, if `a.proto` imports `b.proto`, `protoc-gen-go` needs to know which Go package both `a.pb.go` and `b.pb.go` are in. There's no reliable way to derive the Go import path from the information in a `.proto` file, so you need to tell the compiler what it is.

If you don't include a `option go_package`, then the compiler will try to guess what the import path might be, using a collection of heuristics which have accumulated over time. It might even get it right! But you're probably still better off being explicit.

**Comment by aronatkins** [comment]
Thanks for confirming that `go_package` is the right path forward. I've got it working here. I understand why the proto/Go package mapping is necessary.

I wasn't sure if there was any magical "oh. `a.proto` is in the same package as `b.proto` and only certain files need `go_package`".

**Comment by aronatkins** [comment]
Filed https://github.com/tensorflow/tensorflow/issues/34100 to ask for `go_package` in TF.
